py_zipkin/request_helpers.py

```diff
@@ -0,0 +1,147 @@
+# -*- coding: utf-8 -*-
+import logging
+
+from py_zipkin.util import _should_sample
+from py_zipkin.zipkin import create_attrs_for_span
+from py_zipkin.zipkin import ZipkinAttrs
+
+log = logging.getLogger(__name__)
+
+
+def _parse_single_header(b3_header):
+    """
+    Parse out and return the data necessary for generating ZipkinAttrs.
+
+    Returns a dict with the following keys:
+        'trace_id':             str or None
+        'span_id':              str or None
+        'parent_span_id':       str or None
+        'sampled_str':          '0', '1', 'd', or None
+        'sample_decision_only': True if header contains only a sampling
+                                decision, false-y value otherwise
+    """
+    parsed = dict.fromkeys(('trace_id', 'span_id', 'parent_span_id',
+                            'sampled_str', 'sample_decision_only'))
+
+    # b3={TraceId}-{SpanId}-{SamplingState}-{ParentSpanId}
+    #      (last 2 fields optional)
+    # OR
+    # b3={SamplingState}
+    bits = b3_header.split('-')
+
+    # Handle the lone-sampling-decision case:
+    if len(bits) == 1:
+        parsed['sampled_str'] = bits[0]
+        parsed['sample_decision_only'] = True
+        return parsed
+    if len(bits) > 4:
+        # Too many segments, just return the initialized defaults which will
+        # translate to a None result (no valid incoming span context or
+        # sampling decision)
+        return parsed
+    if len(bits) > 3:
+        parsed['parent_span_id'] = bits[3]
+    if len(bits) > 2:
+        parsed['sampled_str'] = bits[2]
+    if len(bits) > 1:
+        parsed['span_id'] = bits[1]
+        parsed['trace_id'] = bits[0]
+    return parsed
+
+
+def _parse_multi_header(headers):
+    """
+    Parse out and return the data necessary for generating ZipkinAttrs.
+
+    Returns a dict with the following keys:
+        'trace_id':             str or None
+        'span_id':              str or None
+        'parent_span_id':       str or None
+        'sampled_str':          '0', '1', 'd', or None
+        'sample_decision_only': True if header contains only a sampling
+                                decision, false-y value otherwise
+    """
+    parsed = {
+        'trace_id': headers.get('X-B3-TraceId', None),
+        'span_id': headers.get('X-B3-SpanId', None),
+        'parent_span_id': headers.get('X-B3-ParentSpanId', None),
+        'sampled_str': headers.get('X-B3-Sampled', None),
+        'sample_decision_only': False,
+    }
+    if headers.get('X-B3-Flags') == '1':
+        parsed['sampled_str'] = 'd'
+    # Unlike the single-header, "absent" and "present-but-empty-string" are
+    # both representable, and the spec only mentions Absent for encoding
+    # "Defer".  So that means we need to translate an empty-string to a
+    # non-empty, but invalid string.
+    if parsed['sampled_str'] == '':
+        parsed['sampled_str'] = 'INVALID'
```

Comment:
why not use exception handling or just quit out.. this is corrupt.

Reply:
...because I didn't think of it.  But that's a great idea!  I'll have the parsers use exceptions to bail on invalid input and that might simplify the common post-parsing logic.